### PR TITLE
Implement automatic Test Visibility instrumentation for .NET projects

### PR DESCRIPTION
**Pull request by nikita-tkachenko-datadog**
### Requirements for Contributing to this repository

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix one issue at the time.
* The pull request must update the test suite to demonstrate the changed functionality.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see [CONTRIBUTING](/CONTRIBUTING.md). 

### What does this PR do?

Implements automatic APM tracer injection and Test Visibility configuration for .NET projects.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: src/main/java/org/datadog/jenkins/plugins/datadog/apm/DotnetConfigurator.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        String dotnetVersion = workspacePath.act(new ShellCommandCallable(Collections.emptyMap(), GET_DOTNET_VERSION_TIMEOUT_MILLIS, "dotnet", "--version"));
+        listener.getLogger().println("[datadog] Configuring DD .NET tracer: got .NET version " + dotnetVersion + " from " + workspacePath + " on " + node);
+
+        String installTracerOutput = workspacePath.act(new ShellCommandCallable(Collections.emptyMap(), INSTALL_TRACER_TIMEOUT_MILLIS, "dotnet", "tool", "update", "--tool-path", workspacePath.getRemote(), "dd-trace"));
+        listener.getLogger().println("[datadog] Configuring DD .NET tracer: tracer installed in " + workspacePath + " on " + node + "; output: " + installTracerOutput);
+
+        String tracerVarOutput = workspacePath.act(new ShellCommandCallable(Collections.emptyMap(), SHOW_TRACER_VARS_TIMEOUT_MILLIS, workspacePath.getRemote() + File.separator + "dd-trace", "ci", "configure", "jenkins"));
+
+        Map<String, String> variables = new HashMap<>();
+        for (String line : tracerVarOutput.split("\n")) {
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/apm/DotnetConfigurator.java:27]: Can `tracerVarOutput` be `null`?

File: src/main/java/org/datadog/jenkins/plugins/datadog/apm/DotnetConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            if (!line.contains("=")) {
+                continue;
+            }
+            String[] tokens = line.split("=");
+            variables.put(tokens[0], tokens.length == 2 ? tokens[1] : "");
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/apm/DotnetConfigurator.java:32]: What happens if the variables.get(key) returns empty string. Is this expected from the DotNet configuration perspective?

**Review by drodriguezhdez — APPROVED** [review]
I haven't investigated if the actual commands we're using in the configurator are the correct ones, but code wise I've dropped a couple of minor comments.

Apart from that, LGTM.

File: src/main/java/org/datadog/jenkins/plugins/datadog/apm/DotnetConfigurator.java (hunk 2 above)
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/apm/DotnetConfigurator.java:32]: Yes, some of the variables can (maybe even should) have an empty value, checked this with an actual .NET project

File: src/main/java/org/datadog/jenkins/plugins/datadog/apm/DotnetConfigurator.java (hunk 1 above)
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/apm/DotnetConfigurator.java:27]: That'd be an exceptional situation, so a `NullPointerException` would be appropriate (the job itself won't fail anyway, it's just that we'll see the error in the logs)

**Pull request merged**